Wangscape/tilegen/CartesianPowerIterator.h

```diff
@@ -0,0 +1,161 @@
+#pragma once
+#include <vector>
+#include <iterator>
+#include <type_traits>
+#include <algorithm>
+#include <utility>
+#include "common.h"
+
+namespace tilegen
+{
+template<typename InputIt>
+class CartesianPowerIterator : public std::iterator<std::input_iterator_tag,
+                                                    std::vector<typename InputIt::value_type>>
+{
+public:
+    static_assert(std::is_base_of<std::input_iterator_tag,
+                                  typename std::iterator_traits<InputIt>::iterator_category>::value,
+                  "InputIt must be at least an InputIterator");
+    typedef std::vector<typename InputIt::value_type> Values;
+    typedef std::vector<InputIt> Iterators;
+
+    CartesianPowerIterator(InputIt first_, InputIt last_, InputIt init, size_t power);
+    CartesianPowerIterator& operator++();
+
+    bool operator==(const CartesianPowerIterator& other) const;
+    bool operator!=(const CartesianPowerIterator& other) const;
+
+    const Values& operator*() const;
+    const typename InputIt::value_type& get(size_t n) const;
+    const Iterators& getIterators() const;
+    const InputIt& getIterator(size_t n) const;
+    const typename InputIt::difference_type coordinate(size_t n) const;
+    const InputIt& getFirst() const;
+    const InputIt& getLast() const;
+    std::pair<size_t, size_t> coordinates_2d() const;
+private:
+    const InputIt mFirst;
+    const InputIt mLast;
+    std::vector<InputIt> mIterators;
+    std::vector<typename InputIt::value_type> mValues;
+};
+
+template<typename InputIt>
+inline CartesianPowerIterator<InputIt>::CartesianPowerIterator(InputIt first, InputIt last, InputIt init, size_t power) :
+    mFirst(first), mLast(last),
+    mIterators(power, init)
+{
+    if (init != last)
+        for (const auto& it : mIterators)
+            mValues.push_back(*it);
```

Comment:
An initial draft just stored the iterators, and had a function wrapping `std::transform` to output the vector of values. However, this meant needlessly copying all the values every time they were accessed. On the majority of iterations, only one value changes.

With a vector of copies of the values, we make on average one copy per iteration, and the client code can make the decision to copy the whole vector if necessary.